engine/access/state_stream/backend/event_provider.go

```diff
@@ -0,0 +1,175 @@
+package backend
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/access"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/optimistic_sync"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// EventsResponse represents the response containing events for a specific block.
+type EventsResponse struct {
+	BlockID          flow.Identifier
+	Height           uint64
+	Events           flow.EventsList
+	BlockTimestamp   time.Time
+	ExecutorMetadata access.ExecutorMetadata
+}
+
+// EventsProvider retrieves events by block height. It can be configured to retrieve events from
+// the events indexer(if available) or using a dedicated callback to query it from other sources.
+type EventsProvider struct {
+	log              zerolog.Logger
+	headers          storage.Headers
+	execDataProvider ExecutionDataProvider
+
+	fetchFromLocalCache bool
+	execResultProvider  optimistic_sync.ExecutionResultInfoProvider
+	execStateCache      optimistic_sync.ExecutionStateCache
+}
+
+func NewEventsProvider(
+	log zerolog.Logger,
+	headers storage.Headers,
+	execDataProvider ExecutionDataProvider,
+	fetchFromLocalCache bool,
+	execResultProvider optimistic_sync.ExecutionResultInfoProvider,
+	execStateCache optimistic_sync.ExecutionStateCache,
+) *EventsProvider {
+	return &EventsProvider{
+		log:                 log,
+		headers:             headers,
+		execDataProvider:    execDataProvider,
+		fetchFromLocalCache: fetchFromLocalCache,
+		execResultProvider:  execResultProvider,
+		execStateCache:      execStateCache,
+	}
+}
+
+// Events return events for a given block height.
+//
+// Expected error returns during normal operations:
+//   - [codes.NotFound]: If a block header for the specified block height is not found.
+func (b *EventsProvider) Events(
+	ctx context.Context,
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	var response *EventsResponse
+	var err error
+
+	if b.fetchFromLocalCache {
+		response, err = b.getEventsFromStorage(height, criteria)
+	} else {
+		response, err = b.getEventsFromExecutionData(ctx, height, criteria)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get events: %w", err)
+	}
+
+	header, err := b.headers.ByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
```

Comment:
I think you could make this logic a little simpler by refactoring a bit. currently `getEventsFromStorage` and `getEventsFromExecutionData` take a height, then lookup the block id. then this function looks up the block header.

You could instead lookup the header first, then pass the header (or blockID) into the methods.